python/lsst/daf/butler/direct_query_driver/_query_builder.py

```diff
@@ -0,0 +1,573 @@
+
+from __future__ import annotations
+
+__all__ = ("QueryJoiner", "QueryBuilder")
+
+import dataclasses
+import itertools
+from collections.abc import Iterable, Sequence
+from typing import TYPE_CHECKING, Any, ClassVar
+
+import sqlalchemy
+
+from .. import ddl
+from ..name_shrinker import NameShrinker
+from ..nonempty_mapping import NonemptyMapping
+from ..queries import tree as qt
+from ._postprocessing import Postprocessing
+
+if TYPE_CHECKING:
+    from ..registry.interfaces import Database
+    from ..timespan_database_representation import TimespanDatabaseRepresentation
+
+
+@dataclasses.dataclass
+class QueryBuilder:
+    """A struct used to represent an under-construction SQL SELECT query.
+
+    This object's methods frequently "consume" ``self``, by either returning
+    it after modification or returning related copy that may share state with
+    the original.  Users should be careful never to use consumed instances, and
+    are recommended to reuse the same variable name to make that hard to do
+    accidentally.
+    """
+
+    joiner: QueryJoiner
+    """Struct representing the SQL FROM and WHERE clauses, as well as the
+    columns *available* to the query (but not necessarily in the SELECT
+    clause).
+    """
+
+    columns: qt.ColumnSet
+    """Columns to include the SELECT clause.
+
+    This does not include columns required only by `postprocessing` and columns
+    in `QueryJoiner.special`, which are also always included in the SELECT
+    clause.
+    """
+
+    postprocessing: Postprocessing = dataclasses.field(default_factory=Postprocessing)
+    """Postprocessing that will be needed in Python after the SQL query has
+    been executed.
+    """
+
+    distinct: bool | Sequence[sqlalchemy.ColumnElement[Any]] = ()
+    """A representation of a DISTINCT or DISTINCT ON clause.
+
+    If `True`, this represents a SELECT DISTINCT.  If a non-empty sequence,
+    this represents a SELECT DISTINCT ON.  If `False` or an empty sequence,
+    there is no DISTINCT clause.
+    """
+
+    group_by: Sequence[sqlalchemy.ColumnElement[Any]] = ()
+    """A representation of a GROUP BY clause.
+
+    If not-empty, a GROUP BY clause with these columns is added.  This
+    generally requires that every `sqlalchemy.ColumnElement` held in the nested
+    `joiner` that is part of `columns` must either be part of `group_by` or
+    hold an aggregate function.
+    """
+
+    EMPTY_COLUMNS_NAME: ClassVar[str] = "IGNORED"
+    """Name of the column added to a SQL SELECT clause in order to construct
+    queries that have no real columns.
+    """
+
+    EMPTY_COLUMNS_TYPE: ClassVar[type] = sqlalchemy.Boolean
+    """Type of the column added to a SQL SELECT clause in order to construct
+    queries that have no real columns.
+    """
+
+    def __post_init__(self) -> None:
+        assert not (self.distinct and self.group_by), "At most one of distinct and group_by can be set."
```

Comment:
This check should probably be in `select()` instead of post_init... most of the assignments to distinct and group_by occur after initialization.